node/actors/bft/src/leader/tests.rs

```diff
@@ -1,39 +1,398 @@
+
+    let junk = util.rng().gen::<CommitQC>();
+    let replica_prepare = util.new_current_replica_prepare(|msg| msg.high_qc = junk);
+    let res = util.dispatch_replica_prepare(replica_prepare);
+    assert_matches!(res, Err(ReplicaPrepareError::InvalidHighQC(..)));
+}
+
+#[tokio::test]
+async fn replica_prepare_high_qc_of_current_view() {
+    let mut util = UTHarness::new().await;
+
+    let view = ViewNumber(1);
+    let qc_view = ViewNumber(1);
+
+    util.set_view(view);
+    let qc = util.new_commit_qc(|msg| {
+        msg.view = qc_view;
+    });
+    let replica_prepare = util.new_current_replica_prepare(|msg| msg.high_qc = qc);
+    let res = util.dispatch_replica_prepare(replica_prepare);
+    assert_matches!(
+        res,
+        Err(ReplicaPrepareError::HighQCOfFutureView { high_qc_view, current_view }) => {
+            assert_eq!(high_qc_view, qc_view);
+            assert_eq!(current_view, view);
+        }
+    );
+}
+
+#[tokio::test]
+async fn replica_prepare_high_qc_of_future_view() {
+    let mut util = UTHarness::new().await;
+
+    let view = ViewNumber(1);
+    let qc_view = ViewNumber(2);
+
+    util.set_view(view);
+    let qc = util.new_commit_qc(|msg| {
+        msg.view = qc_view;
+    });
+
+    let replica_prepare = util.new_current_replica_prepare(|msg| msg.high_qc = qc);
+    let res = util.dispatch_replica_prepare(replica_prepare);
+    assert_matches!(
+        res,
+        Err(ReplicaPrepareError::HighQCOfFutureView{ high_qc_view, current_view }) => {
+            assert_eq!(high_qc_view, qc_view);
+            assert_eq!(current_view, view);
+        }
+    );
+}
+
+#[ignore = "fails/unsupported"]
+#[tokio::test]
+async fn replica_prepare_non_validator_signer() {
+    let mut util = UTHarness::new_with(2).await;
+
+    let view = ViewNumber(2);
+    util.set_view(view);
+    assert_eq!(util.view_leader(view), util.key_at(0).public());
+
+    let replica_prepare = util.new_current_replica_prepare(|_| {});
+    let _ = util.dispatch_replica_prepare(replica_prepare.clone());
+
+    let non_validator: validator::SecretKey = util.rng().gen();
+    let replica_prepare = non_validator.sign_msg(replica_prepare.msg);
+    util.dispatch_replica_prepare(replica_prepare).unwrap();
+    // PANICS:
+    // "Couldn't create justification from valid replica messages!: Message signer isn't in the validator set"
+}
+
+#[tokio::test]
+async fn replica_commit_sanity() {
+    let mut util = UTHarness::new().await;
+
+    let leader_prepare = util.new_procedural_replica_commit().await;
+    util.dispatch_replica_commit(leader_prepare).unwrap();
+}
+
+#[tokio::test]
+async fn replica_commit_sanity_yield_leader_commit() {
+    let mut util = UTHarness::new().await;
+
+    let replica_commit = util.new_procedural_replica_commit().await;
+    util.dispatch_replica_commit(replica_commit).unwrap();
+    util.recv_signed()
+        .await
+        .unwrap()
+        .cast::<LeaderCommit>()
+        .unwrap();
+}
```

Comment:
Sanity tests need to use more replicas. Using just one replica doesn't test the threshold well. Imagine the leader only creates a response when all the replicas answer, or when just one replica answer. We need to have 6 replicas at least, see that the leader creates a response with 5 replica messages.

Reply:
I refactored sanity tests of the 4 types to use a larger validator set. Also, added assertions for every distinct msg processing from the set. 

Changes here: https://github.com/matter-labs/era-consensus/pull/39/commits/9cd6e59ef7fb080634d1f86d2d22f65a2c2c8095